ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsFlatStreamReader.scala

```diff
@@ -0,0 +1,243 @@
+import akka.stream.Attributes
+import akka.stream.scaladsl.Source
+import com.daml.ledger.api.TraceIdentifiers
+import com.daml.ledger.api.v1.event.Event
+import com.daml.ledger.api.v1.transaction_service.GetTransactionsResponse
+import com.daml.ledger.offset.Offset
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.{DatabaseMetrics, Metrics, Timed}
+import com.daml.nameof.NameOf.qualifiedNameOfCurrentFunc
+import com.daml.platform.TemplatePartiesFilter
+import com.daml.platform.configuration.TransactionsFlatStreamReaderConfig
+import com.daml.platform.indexer.parallel.BatchN
+import com.daml.platform.store.backend.EventStorageBackend
+import com.daml.platform.store.backend.common.{
+  EventIdSourceForStakeholders,
+  EventPayloadSourceForFlatTx,
+}
+import com.daml.platform.store.dao.{DbDispatcher, EventProjectionProperties, PaginatingAsyncStream}
+import com.daml.platform.store.dao.PaginatingAsyncStream.IdPaginationState
+import com.daml.platform.store.dao.events.EventsTable.TransactionConversions
+import com.daml.platform.store.dao.events.FilterTableACSReader.{
+  IdQueryConfiguration,
+  statefulDeduplicate,
+}
+import com.daml.platform.store.utils.{ConcurrencyLimiter, QueueBasedConcurrencyLimiter, Telemetry}
+import com.daml.telemetry
+import com.daml.telemetry.Spans
+
+import scala.util.chaining._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{ExecutionContext, Future}
+
+class TransactionsFlatStreamReader(
+    config: TransactionsFlatStreamReaderConfig,
+    globalIdQueriesLimiter: ConcurrencyLimiter,
+    globalPayloadQueriesLimiter: ConcurrencyLimiter,
+    dbDispatcher: DbDispatcher,
+    queryNonPruned: QueryNonPruned,
+    eventStorageBackend: EventStorageBackend,
+    lfValueTranslation: LfValueTranslation,
+    metrics: Metrics,
+)(implicit executionContext: ExecutionContext) {
+  import TransactionsReader._
+  import config._
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val dbMetrics = metrics.daml.index.db
+
+  private val orderBySequentialEventId =
+    Ordering.by[EventStorageBackend.Entry[Raw.FlatEvent], Long](_.eventSequentialId)
+
+  def streamFlatTransactions(
+      queryRange: EventsRange[(Offset, Long)],
+      filteringConstraints: TemplatePartiesFilter,
+      eventProjectionProperties: EventProjectionProperties,
+  )(implicit loggingContext: LoggingContext): Source[(Offset, GetTransactionsResponse), NotUsed] = {
+    val startExclusiveOffset: Offset = queryRange.startExclusive._1
+    val endInclusiveOffset: Offset = queryRange.endInclusive._1
+    val span =
+      Telemetry.Transactions.createSpan(startExclusiveOffset, endInclusiveOffset)(
+        qualifiedNameOfCurrentFunc
+      )
+    logger.debug(
+      s"streamFlatTransactions($startExclusiveOffset, $endInclusiveOffset, $filteringConstraints, $eventProjectionProperties)"
+    )
+    val sourceOfFlatTransactions: Source[(Offset, GetTransactionsResponse), NotUsed] =
```

Comment:
why declare it when it is used just once?

Reply:
Inlined